agenthub/self_discover_agent/agent.py

```diff
@@ -0,0 +1,133 @@
+from agenthub.codeact_agent import CodeActAgent
+from opendevin.controller.agent import Agent
+from opendevin.controller.state.state import State
+from opendevin.core.config import AgentConfig
+from opendevin.core.message import Message, TextContent
+from opendevin.events.action import (
+    Action,
+    AgentFinishAction,
+)
+from opendevin.llm.llm import LLM
+from opendevin.runtime.plugins import PluginRequirement
+
+from .action_parser import SelfDiscoverResponseToActionParser
+from .agent_state_machine import SelfDiscoverState, SelfDiscoverStateMachine
+from .prompt import (
+    IMPLEMENT_EXAMPLE,
+    SYSTEM_MESSAGE,
+    TASK_KEY,
+    get_prompt,
+)
+from .reasoning_action import ReasoningAction
+from .reasoning_action_parser import SelfDiscoverResponseToReasoningActionParser
+
+
+class SelfDiscoverAgent(Agent):
+    VERSION = '0.1'
+    """
+    This agent implements a Self Discover Agent as described in https://arxiv.org/abs/2402.03620.
+    The reasoning structure is sent to the CodeActAgent for execution.
+    """
+
+    action_parser = SelfDiscoverResponseToActionParser()
+    reasoning_parser = SelfDiscoverResponseToReasoningActionParser()
+    sandbox_plugins: list[PluginRequirement] = CodeActAgent.sandbox_plugins
+    system_message = SYSTEM_MESSAGE
+    implement_example = IMPLEMENT_EXAMPLE.format(
+        implement_state_key=SelfDiscoverState.IMPLEMENT.value,
+        task_key=TASK_KEY,
+    )
```

Comment:
It will be nice if we can use the newly implemented Jinja template for this agent - but no pressure - we can do it follow up PR.

https://github.com/OpenDevin/OpenDevin/blob/14a4e45cbbbb01eba01b47e5d874c2583decdfd8/agenthub/codeact_agent/codeact_agent.py#L76-L80

If the template is too rigid, we can figure out a way to generalize it a bit

Reply:
Okay, given your comment below let's defer to a future PR.